File: internal/prompts/validation.go
package prompts

import (
	"fmt"
	"regexp"
	"strings"
)

// Validation functions for user input

// ValidateProjectName validates the project name
func ValidateProjectName(name string) error {
	if name == "" {
		return fmt.Errorf("project name cannot be empty")
	}
	
	// Check length
	if len(name) < 2 {
		return fmt.Errorf("project name must be at least 2 characters")
	}
	if len(name) > 50 {
		return fmt.Errorf("project name must be less than 50 characters")
	}
	
	// Check format (letters, numbers, hyphens, underscores)
	validName := regexp.MustCompile(`^[a-zA-Z][a-zA-Z0-9_-]*$`)
	if !validName.MatchString(name) {
		return fmt.Errorf("project name must start with a letter and contain only letters, numbers, hyphens, and underscores")
	}
	
	// Check for reserved names
	reserved := []string{"test", "main", "init", "vendor", "internal"}
	for _, r := range reserved {
		if strings.EqualFold(name, r) {
			return fmt.Errorf("'%s' is a reserved name", name)
		}
	}
	
	return nil
}

// ValidateModulePath validates the Go module path
func ValidateModulePath(path string) error {
	if path == "" {
		return fmt.Errorf("module path cannot be empty")
	}
	
	// Basic module path validation
	// Should be something like: github.com/user/project
	parts := strings.Split(path, "/")
	if len(parts) < 2 {
		return fmt.Errorf("module path should include at least domain and project (e.g., github.com/user/project)")
	}
	
	// Check for common patterns
	validModule := regexp.MustCompile(`^[a-zA-Z0-9.-]+(/[a-zA-Z0-9._-]+)+$`)
	if !validModule.MatchString(path) {
		return fmt.Errorf("invalid module path format")
	}
	
	return nil
}

// ValidateGoVersion is defined in go_version.go

File: web/src/utils/validation.ts
// Common validation patterns
export const ValidationPatterns = {
  // Project name: letters, numbers, hyphens, underscores
  projectName: /^[a-zA-Z0-9-_]+$/,
  
  // Go module path: domain/user/project format
  goModule: /^[a-zA-Z0-9.-]+\/[a-zA-Z0-9._\/-]+$/,
  
  // Semantic version
  version: /^\d+\.\d+(\.\d+)?$/,
  
  // Email
  email: /^[^\s@]+@[^\s@]+\.[^\s@]+$/,
  
  // URL
  url: /^https?:\/\/(www\.)?[-a-zA-Z0-9@:%._\+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b([-a-zA-Z0-9()@:%_\+.~#?&//=]*)$/,
}

// Validation error messages
export const ValidationMessages = {
  projectName: {
    required: 'Project name is required',
    pattern: 'Use only letters, numbers, hyphens, and underscores',
    minLength: 'Project name must be at least 2 characters',
    maxLength: 'Project name must be less than 50 characters',
  },
  goModule: {
    required: 'Module path is required',
    pattern: 'Must be a valid Go module path (e.g., github.com/user/project)',
    example: 'Example: github.com/username/project-name',
  },
  version: {
    pattern: 'Must be a valid version (e.g., 1.21)',
  },
}

// Custom validation functions
export const CustomValidators = {
  projectName: (value: string): string | null => {
    if (!value) return ValidationMessages.projectName.required
    if (!ValidationPatterns.projectName.test(value)) return ValidationMessages.projectName.pattern
    if (value.length < 2) return ValidationMessages.projectName.minLength
    if (value.length > 50) return ValidationMessages.projectName.maxLength
    return null
  },
  
  goModule: (value: string): string | null => {
    if (!value) return ValidationMessages.goModule.required
    if (!ValidationPatterns.goModule.test(value)) {
      return `${ValidationMessages.goModule.pattern}. ${ValidationMessages.goModule.example}`
    }
    return null
  },
  
  uniqueName: (existingNames: string[]) => (value: string): string | null => {
    if (existingNames.includes(value)) {
      return 'This name is already taken'
    }
    return null
  },
}